kernel/Dockerfile

```diff
@@ -28,12 +28,16 @@ RUN apk add \
     xz-dev \
     zlib-dev
 
+# Update patch utility
+RUN apk update && apk add --update patch
```

Comment:
Why is this needed?

Reply:
When I applied one of those -rt patches,  I cannot built that successfully.

net/core/dev.o: In function `net_rx_action':
/linux/net/core/dev.c:5658: undefined reference to `__raise_softirq_irqoff_ksoft'
net/core/dev.o:(.debug_addr+0x6e50): undefined reference to `__raise_softirq_irqoff_ksoft'

One -rt patch, patches-4.14.x-rt/net-provide-a-way-to-delegate-processing-a-softirq-t.patch, is trying to refine __raise_softirq_irqoff_ksoft in kernel/softirq.c. But there are two similar contexts isolated with #-rt #else #endif by one previous -rt patch. The original `patch` command just puts __raise_softirq_irqoff_ksoft into {#else #endif} so I have to update `patch`.

Reply:
If patch is doing it wrong then that suggests that the patches need refreshing. For automated things like this using `-F 0` to turn off fuzzing and require patches to apply cleanly should be safer than relying on the quircks of a particular patch implementation to DTRT.

Reply:
I can do this but why can the updated patch utility apply that correctly? So I prefer to update patch utility, instead. This kind of thing could happen again in the future.

Reply:
Applying with fuzz involves applying heuristics which can differ across implementations (or even versions of the same implementation). Using `-F 0` disables fuzzing and doesn't have that issue.

Reply:
FWIW `dpgk-source` seems to use `-t -F0 -N -p1 -u` which might be good to apply throughout `kernel/*` not just here.

Reply:
@ijc Let make this clear :) The original patch is one simple version from busybox so it doesn't support -F 0. I can show you as follows:

...
Patching -rt kernel
patch: unrecognized option: F
BusyBox v1.27.2 (2017-12-12 10:41:50 GMT) multi-call binary.

Usage: patch [OPTIONS] [ORIGFILE [PATCHFILE]]

        -p,--strip N            Strip N leading components from file names
        -i,--input DIFF         Read DIFF instead of stdin
        -R,--reverse            Reverse patch
        -N,--forward            Ignore already applied patches
        -E,--remove-empty-files Remove output files if they become empty

Reply:
Can you then add `patch` to the list of utilities installed just above where you add it?

Reply:
That's a shame but the fact remains that the solution to "patch with fuzz is applied incorrectly" is to fix the patch not to change the implementation of patch being used.

There's certainly an argument for switch to a fuller patch so that we can use `-F0` but that would not solve the issue you are having here with the fuzzy patch.

Reply:
Hmm, I thought I commented on this. Anyway, I agree with @ijc that the patches should apply without fuzz. For the other patches I maintain a linux-stable tree and rebase the branches on every kernel update, so it's guaranteed. I prefer if the `-rt` patches are the same as I don't really want to deal with patch conflicts.

Reply:
@ijc @rn I double checked this. This is not sort of fuzzy patch. In that patch, this change is,
+++ b/kernel/softirq.c
@@ -686,6 +686,27 @@ void __raise_softirq_irqoff(unsigned int
 }

 /*
+ * Same as __raise_softirq_irqoff() but will process them in ksoftirqd
+ */
+void __raise_softirq_irqoff_ksoft(unsigned int nr)
...

686 is exact and correct line in kernel/softirq.c as expected. But that original patch always applies this in wrong place with wrong line and even wrong context. Is this one bug to the original patch?

Reply:
It does indeed sound like a bug in `busybox patch` (possibly one worth reporting upstream).

In the end it's a bit moot since @rn's change in #2909 (d0d7738efac6127263827f60b65e22eb34264ee4) to use full patch with `-F0` and friends. If you rebase this whole hunk/discussion should go away.

Reply:
@ijc Good! Now it works for me with @rn's change #2909.